File: 23.tell-dont-ask/src/java/src/test/java/com/telldontask/doubles/InMemoryProductCatalog.java
package com.telldontask.doubles;

import com.telldontask.domain.Product;
import com.telldontask.repository.ProductCatalog;

import java.util.List;

public class InMemoryProductCatalog implements ProductCatalog {
    private final List<Product> products;

    public InMemoryProductCatalog(List<Product> products) {
        this.products = products;
    }

    @Override
    public Product getByName(String name) {
        return products.stream()
                .filter(p -> p.getName().equals(name))
                .findFirst()
                .orElse(null);
    }
}

File: 23.tell-dont-ask/src/typescript/tests/doubles/inMemoryProductCatalog.ts
import { Product } from '../../domain/product';
import { ProductCatalog } from '../../repository/productCatalog';

export class InMemoryProductCatalog implements ProductCatalog {
  constructor(private readonly products: Product[]) {}

  getByName(name: string): Product | undefined {
    return this.products.find((p) => p.name === name);
  }
}
